amethyst_animation/src/systems/sampling.rs

```diff
@@ -0,0 +1,189 @@
+use std::time::{Duration, Instant};
+
+use amethyst_assets::AssetStorage;
+use amethyst_core::{duration_to_secs, LocalTransform};
+use specs::{Fetch, Join, System, WriteStorage};
+
+use interpolation::Interpolate;
+use resources::{ControlState, EndControl, RestState, Sampler, SamplerControl, SamplerControlSet};
+
+/// System for interpolating active samplers.
+///
+/// If other forms of animation is needed, this can be used in isolation, have no direct dependency
+/// on `AnimationControlSystem`.
+#[derive(Default)]
+pub struct SamplerInterpolationSystem;
+
+impl SamplerInterpolationSystem {
+    pub fn new() -> Self {
+        Default::default()
+    }
+}
+
+impl<'a> System<'a> for SamplerInterpolationSystem {
+    type SystemData = (
+        Fetch<'a, AssetStorage<Sampler>>,
+        WriteStorage<'a, SamplerControlSet>,
+        WriteStorage<'a, LocalTransform>,
+    );
+
+    fn run(&mut self, (samplers, mut controls, mut transforms): Self::SystemData) {
+        let now = Instant::now();
+        for (control_set, transform) in (&mut controls, &mut transforms).join() {
+            match control_set.translation {
```

Comment:
could be `if let Some(sampler) = control_set.translation.as_mut().and_then(|c| samplers.get(c.sampler))`